routing/edge_estimator.cpp

```diff
@@ -27,54 +38,90 @@ double CalcTrafficFactor(SpeedGroup speedGroup)
 
-  double const distanceM = MercatorBounds::DistanceOnEarth(from, to);
-  return distanceM / speedMPS;
+  double const speedMPS = road.GetSpeed() * kKMPH2MPS;
+  double result = TimeBetweenSec(road.GetPoint(segment.GetPointId(false /* front */)),
+                                 road.GetPoint(segment.GetPointId(true /* front */)), speedMPS);
+
+  return result;
+}
+
+double EdgeEstimator::CalcHeuristic(m2::PointD const & from, m2::PointD const & to) const
+{
+  return TimeBetweenSec(from, to, m_maxSpeedMPS);
+}
+
+double EdgeEstimator::CalcLeapWeight(m2::PointD const & from, m2::PointD const & to) const
+{
+  // Let us assume for the time being that
+  // leap edges will be added with a half of max speed.
+  // @TODO(bykoianko) It's necessary to gather statistics to calculate better factor(s) instead of
+  // one below.
+  return TimeBetweenSec(from, to, m_maxSpeedMPS / 2.0);
 }
 
-class CarEdgeEstimator : public EdgeEstimator
+// PedestrianEstimator -----------------------------------------------------------------------------
+class PedestrianEstimator final : public EdgeEstimator
 {
 public:
-  CarEdgeEstimator(shared_ptr<TrafficStash> trafficStash, double maxSpeedKMpH);
+  explicit PedestrianEstimator(double maxSpeedKMpH) : EdgeEstimator(maxSpeedKMpH) {}
+
+  // EdgeEstimator overrides:
+  double GetUTurnPenalty() const override { return 0.0 /* seconds */; }
+  bool LeapIsAllowed(NumMwmId /* mwmId */) const override { return false; }
+};
+
+// BicycleEstimator --------------------------------------------------------------------------------
+class BicycleEstimator final : public EdgeEstimator
+{
+public:
+  explicit BicycleEstimator(double maxSpeedKMpH) : EdgeEstimator(maxSpeedKMpH) {}
+
+  // EdgeEstimator overrides:
+  double GetUTurnPenalty() const override { return 20.0 /* seconds */; }
+  bool LeapIsAllowed(NumMwmId /* mwmId */) const override { return false; }
+};
+
+// CarEstimator ------------------------------------------------------------------------------------
+class CarEstimator final : public EdgeEstimator
+{
+public:
+  CarEstimator(shared_ptr<TrafficStash> trafficStash, double maxSpeedKMpH);
 
   // EdgeEstimator overrides:
   double CalcSegmentWeight(Segment const & segment, RoadGeometry const & road) const override;
-  double CalcHeuristic(m2::PointD const & from, m2::PointD const & to) const override;
-  double CalcLeapWeight(m2::PointD const & from, m2::PointD const & to) const override;
   double GetUTurnPenalty() const override;
   bool LeapIsAllowed(NumMwmId mwmId) const override;
 
 private:
   shared_ptr<TrafficStash> m_trafficStash;
-  double const m_maxSpeedMPS;
 };
 
-CarEdgeEstimator::CarEdgeEstimator(shared_ptr<TrafficStash> trafficStash, double maxSpeedKMpH)
-  : m_trafficStash(trafficStash), m_maxSpeedMPS(maxSpeedKMpH * kKMPH2MPS)
+CarEstimator::CarEstimator(shared_ptr<TrafficStash> trafficStash, double maxSpeedKMpH)
+  : EdgeEstimator(maxSpeedKMpH)
+  , m_trafficStash(move(trafficStash))
```

Comment:
This is an attempt to save some clocks on atomic ops, but from my point of view even without move the code is fast enough, and I believe that optimizer will do the job.